crates/network-monitor/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::collections::VecDeque;
+use std::fs;
+use std::future::Future;
+use std::io;
+use std::mem;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use camino::*;
+use chrono::{DateTime, TimeDelta, Utc};
+use serde::*;
+use thiserror::Error;
+use tokio::select;
+use tokio_util::sync::CancellationToken;
+
+use network_scanner::ping;
+
+mod log_queue;
+mod state;
+
+pub use crate::state::State;
+
+#[derive(Error, Debug)]
+#[error(transparent)]
+pub enum SetConfigError {
+    Io(#[from] io::Error),
+    Serde(#[from] serde_json::Error),
+}
+
+pub async fn set_config(config: MonitorsConfig, state: Arc<State>) -> Result<(), SetConfigError> {
+    let file = fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(get_monitor_config_path(&state.cache_path))?;
+
+    let mut config_write = state.config.write().await;
+
+    serde_json::to_writer_pretty(&file, &config)?;
+
+    let old_config = mem::replace(&mut *config_write, config);
+
+    let new_config_set: HashSet<&MonitorDefinition> = config_write.monitors.iter().collect();
+    let old_config_set: HashSet<&MonitorDefinition> = old_config.monitors.iter().collect();
+
+    let added = new_config_set.difference(&old_config_set);
+    let deleted = old_config_set.difference(&new_config_set);
+
+    let (new_cancellation_tokens, new_monitors): (
+        Vec<(String, CancellationToken)>,
+        Vec<Pin<Box<dyn Future<Output = ()> + Send>>>,
+    ) = added
+        .map(|definition| {
+            let cancellation_token = CancellationToken::new();
+            let cancellation_monitor = cancellation_token.clone();
+
+            let definition_clone = (*definition).clone(); // TODO: is there a nicer way to do this?
```

Comment:
If I remember right, I have to clone `definition` here because its owner will not outlive the end of the function. But I suspect there might be a way to do `map` so that `definition` is already a clone?

Reply:
Thanks, I got it working. I had missed that you were using the clone**d** function, which is the key here